From 39585b79c1c09a92b6ac4c1552d38d345b9637d9 Mon Sep 17 00:00:00 2001
From: Robert Dolca <robert.dolca@intel.com>
Date: Tue, 25 Aug 2015 21:15:19 +0300
Subject: [PATCH 13/17] input: silead: activate in-kernel slot tracking

The silead touchscreens don't support slot ids (it is always 0).
Use input_mt_assign_slots to match the set of touches against the
currently used slots via the Lagrange relaxation algorithm.

Change-Id: Ia5d057bc25d332d7629c67ae75c7c29e76972657
Tracked-On: https://jira01.devtools.intel.com/browse/OAM-20919
Signed-off-by: Robert Dolca <robert.dolca@intel.com>
Signed-off-by: Constantin Musca <constantin.musca@intel.com>
Reviewed-on: https://android.intel.com:443/472786
---
 drivers/input/touchscreen/silead.c | 21 ++++++++++++++++++---
 1 file changed, 18 insertions(+), 3 deletions(-)

diff --git a/drivers/input/touchscreen/silead.c b/drivers/input/touchscreen/silead.c
index dc8a352..94dd3fb 100644
--- a/drivers/input/touchscreen/silead.c
+++ b/drivers/input/touchscreen/silead.c
@@ -97,6 +97,8 @@ struct silead_ts_data {
 	u8 xy_swap;
 	u8 x_invert;
 	u8 y_invert;
+	struct input_mt_pos pos[SILEAD_MAX_FINGERS];
+	int slots[SILEAD_MAX_FINGERS];
 };
 
 struct silead_fw_data {
@@ -130,7 +132,8 @@ static int silead_ts_request_input_dev(struct silead_ts_data *data)
 			     200, 0, 0);
 
 	input_mt_init_slots(data->input_dev, data->max_fingers,
-			    INPUT_MT_DIRECT | INPUT_MT_DROP_UNUSED);
+			    INPUT_MT_DIRECT | INPUT_MT_DROP_UNUSED |
+			    INPUT_MT_TRACK);
 
 	data->input_dev->name = SILEAD_TS_NAME;
 	data->input_dev->phys = "input/ts";
@@ -202,8 +205,18 @@ static void silead_ts_read_data(struct i2c_client *client)
 
 		x = le16_to_cpup((__le16 *)(buf + offset + SILEAD_POINT_X_OFF));
 		y = le16_to_cpup((__le16 *)(buf + offset + SILEAD_POINT_Y_OFF));
+		data->pos[i - 1].x = x;
+		data->pos[i - 1].y = y;
 
 		dev_dbg(dev, "x=%d y=%d id=%d\n", x, y, id);
+	}
+
+	input_mt_assign_slots(data->input_dev, data->slots, data->pos, touch_nr);
+
+	for (i = 0; i < touch_nr; i++) {
+		x = data->pos[i].x;
+		y = data->pos[i].y;
+		id = data->slots[i];
 
 		if (data->xy_swap)
 			silead_ts_report_touch(data,
@@ -211,14 +224,16 @@ static void silead_ts_read_data(struct i2c_client *client)
 					       data->y_max - y : y,
 					       data->x_invert ?
 					       data->x_max - x : x,
-					       i - 1);
+					       id);
 		else
 			silead_ts_report_touch(data,
 					       data->x_invert ?
 					       data->x_max - x : x,
 					       data->y_invert ?
 					       data->y_max - y : y,
-					       i - 1);
+					       id);
+
+		dev_dbg(dev, "x=%d y=%d sw_id=%d\n", x, y, id);
 	}
 
 	input_mt_sync_frame(data->input_dev);
-- 
1.9.1

